
#Script to analyse the frequency data and associate with features. 
* Read the csv files 
* Perform necessary calcuations
*Plot results (eventually new script)

Make sure to set your working directory to the correct folder
Kristof: setwd('/Users/hopla/project-master/project/gitpleuni/bachelerProject/Rscripts')

```{r}
source('./baseRscript.R')
```

*Read the stored frequencies rather than calculating frequencies again*
```{r}
if (TRUE){#read the stored data
read.table("../Output/freqPatTsInclDay0-threshold.csv",sep=",",header=TRUE,row.names=1)->freqPatTs
}
```

*Calculate mean freq for each site and bootstrap* 
```{r}
#Get mean
#Calculate mean freq for each site 
#attach(freqPatSite); colMeans<-apply(freqPatSite, 2 , mean, na.rm=TRUE)
attach(freqPatTs); colMeansTs<-apply(freqPatTs, 2 , mean, na.rm=TRUE)

#Get conf intervals by bootsstrapping
#PSP Nov 11 2015: I don't think we need te bootstrapping, so I put it in an if (FALSE) statement.
#ts
if (FALSE){
btmeansTs<-data.frame(row.names=names(freqPatTs)[1:984]) #each row is a site, each column is a bootstrapped mean
numbootstraps=1000
for (j in 1:984){# start with the first site
    print(j)
for (k in 1:numbootstraps){ # first iteration
  btmeansTs[j,k]=mean(sample(freqPatTs[,j],length(freqPatTs[,j]),replace = TRUE),na.rm=TRUE)
}}
btmeansSorted<-t(apply(btmeansTs,1,sort))
lowerConf<-btmeansSorted[,floor(0.025*numbootstraps)]
upperConf<-btmeansSorted[,floor((1-0.025)*numbootstraps)]
}
```

## Method 2: original update by  Pleuni 
Only synonymous, non-synomous and stop codons are considered
- for each mutation, determine whether it is synonymous, non-synonymous or creates a stop
- add information on resistance  positions

```{r}
#PSP Nov 11 2015 I removed lowerConf and upperConf here because we no longer calculate them
#x<-data.frame(num=1:984,colMeansTs,TypeOfSite,lowerConf,upperConf)
#PSP Nov 11 2015 I renamed x OverviewDF and newdata OverviewDFOrderedByFreq
OverviewDF<-data.frame(num=1:984,colMeansTs,TypeOfSite)
OverviewDFOrderedByFreq <- OverviewDF[order(colMeansTs),] 
OverviewDFOrderedByFreq$color<-""
for (i in 1:984){
    if (OverviewDFOrderedByFreq$TypeOfSite[i]=="syn") OverviewDFOrderedByFreq$color[i] = "darkolivegreen3"
    if (OverviewDFOrderedByFreq$TypeOfSite[i]=="nonsyn") OverviewDFOrderedByFreq$color[i] = "red"
    if (OverviewDFOrderedByFreq$TypeOfSite[i]=="stop") OverviewDFOrderedByFreq$color[i] = "black"
    if (OverviewDFOrderedByFreq$TypeOfSite[i]=="res") OverviewDFOrderedByFreq$color[i] = "purple"
}

```


```{r}
#for ease of reading 
freqPatTs<-round(freqPatTs,2)
```


# Selecting the Protease or the Reverse transcriptase region 

*PRO: Make the plots (transitions)*

```{r}
wlibrary(scales)
pdf("../Output/Protease.pdf",width = 13, height = 10)
PROdata<-OverviewDFOrderedByFreq[OverviewDFOrderedByFreq$num<298,]
#remove resistance mutations
PROdata<-PROdata[PROdata$TypeOfSite!="res",]
plot(PROdata$colMeansTs+0.001,log="y",ylim=c(0.001,0.15),cex=1.5, pch = 16, col=alpha(PROdata$color, 1), xlab = "Nucleotides ordered by mean mutation frequency", ylab = "Mean mutation frequency" , yaxt = "n")
axis(2,labels = c(0,0.001, 0.005, 0.05, 0.1), at = c(0.001, 0.002, 0.006, 0.051, 0.101))
dev.off()
```

*RT: Make the plots (transitions)*

```{r}
pdf("../Output/RT-version.pdf",width = 13, height = 10)
RTdata<-OverviewDFOrderedByFreq[OverviewDFOrderedByFreq$num>=298,]
RTdata<-RTdata[RTdata$TypeOfSite!="res",]
plot(RTdata$colMeansTs+0.001,log="y",ylim=c(0.001,0.15),cex=1, pch = "|",col=RTdata$color)
#arrows(1:length(RTdata$num), RTdata$lowerConf +0.001, 1:length(RTdata$num), RTdata$upperConf +0.001, 
#       angle=90, length=0.01, code=3, col=RTdata$color,lwd=0.1)
dev.off()

```

### Make an AA variability map 

```{r}
translatedB<-seqinr::translate(consensusB)
#convert the fasta files into AA sequences
translatedfasta<-c()
for (i in 1:length(listfastafiles)){ #for each fastafile 
         filename=paste("../Data/BachelerFiles/FASTAfiles/",substr(listfastafiles[i],1,6),".fasta",sep="")
	       patfasta<-read.dna(filename, format = "fasta",as.character=TRUE) #read the file       
	       translatedfasta<-rbind(translatedfasta,t(apply(patfasta,1,function(x) seqinr::translate(x))))
}
```

```{r}
#determine for each codon position, the frequency of mutations compared to subtype B
mutations<-c()
for(m in 1:length(translatedB)){
prevmut<-100-(length(which(translatedfasta[,m]==translatedB[m]))/nrow(translatedfasta)*100)
mutations<-c(mutations,prevmut)
}

# Plot for protease including resistance positions
plot(mutations[1:99],type='o',ylab='Mutation frequency (%)',xlab='AA position')
# Plot for RT including resistance positions
plot(mutations[ 100:328],type='o',ylab='Mutation frequency (%)',xlab='AA position',xaxt='n')
axis(1,(100:328)-99,label=(100:328)-99)


# Plot for RT excluding resistance positions and including resisatnce positions in red
mutationposition<-cbind(1:229,mutations[100:328])
plot(mutationposition[-c(sort(RTImuts$pos)),1],mutationposition[-c(sort(RTImuts$pos)),2],type='o',ylab='Mutation frequency (%)',xlab='AA position',xaxt='n')
points(mutationposition[c(sort(RTImuts$pos)),1],mutationposition[c(sort(RTImuts$pos)),2],col='red')
axis(1,(100:328)-99,label=(100:328)-99)
```



## Non-Syn plot 

```{r}
pdf("../Output/PR-nonsyn.pdf",width = 13, height = 10)
PROdataNonSyn<-PROdata[PROdata$TypeOfSite=="nonsyn",]
#Make plot of Protease 
plot(PROdataNonSyn$num[order(PROdataNonSyn$num)],PROdataNonSyn$colMeansTs[order(PROdataNonSyn$num)]+0.001,col=PROdataNonSyn$color[order(PROdataNonSyn$num)],pch=16,log="y",type="b")#,ylim=c(0,max(PROdataNonSyn$colMeansTs)))
#Add moving mean
for (i in 20:280){
    range = (i-20):(i+20)
#    points(i, mean(PROdataNonSyn$colMeansTs[PROdataNonSyn$num%in%range]), col="blue",pch = 12, cex=3)
    points(i, median(PROdataNonSyn$colMeansTs[PROdataNonSyn$num%in%range]), col="green",pch = 11, cex=2)
}
dev.off()

RTdataNonSyn<-RTdata[RTdata$TypeOfSite=="nonsyn",]
plot(RTdataNonSyn$num[order(RTdataNonSyn$num)],RTdataNonSyn$colMeansTs[order(RTdataNonSyn$num)]+0.001,col=RTdataNonSyn$color[order(RTdataNonSyn$num)],pch=16,log="y",type="b")

#Add moving mean
for (i in 370:900){
    range = (i-20):(i+20)
   # points(i, mean(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="blue",pch = 12, cex=3)
    points(i, median(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="green",pch = 11, cex=3)
}

i=370
    range = (i-20):(i+20)
```


## Syn plot 

```{r}
pdf("../Output/PR-syn.pdf",width = 13, height = 10)
PROdataSyn<-PROdata[PROdata$TypeOfSite=="syn",]
#Make plot of Protease 
plot(PROdataSyn$num[order(PROdataSyn$num)],PROdataSyn$colMeansTs[order(PROdataSyn$num)]+0.001,col=PROdataSyn$color[order(PROdataSyn$num)],pch=16,
     log="y",
     type="b")#,ylim=c(0,max(PROdataNonSyn$colMeansTs)))

#Add moving mean
for (i in 20:280){
    range = (i-10):(i+10)
    points(i, median(PROdataSyn$colMeansTs[PROdataSyn$num%in%range]), col="blue",pch = 11, cex=2)
}
dev.off()
```

```{r}
RTdataNonSyn<-RTdata[RTdata$TypeOfSite=="nonsyn",]
plot(RTdataNonSyn$num[order(RTdataNonSyn$num)],RTdataNonSyn$colMeansTs[order(RTdataNonSyn$num)]+0.001,col=RTdataNonSyn$color[order(RTdataNonSyn$num)],pch=16,log="y",type="b")

#Add moving mean
for (i in 370:900){
    range = (i-20):(i+20)
   # points(i, mean(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="blue",pch = 12, cex=3)
    points(i, median(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="green",pch = 11, cex=3)
}

i=370
    range = (i-20):(i+20)
```

```{r}
pdf("../Output/RT-syn.pdf",width = 13, height = 10)
RTdataSyn<-RTdata[RTdata$TypeOfSite=="syn",]
#Make plot of RTtease 
plot(RTdataSyn$num[order(RTdataSyn$num)],RTdataSyn$colMeansTs[order(RTdataSyn$num)]+0.001,col=RTdataSyn$color[order(RTdataSyn$num)],pch=16,
     log="y",
     type="b", 
    ylim = c(0.001,0.06)    )#,ylim=c(0,max(RTdataNonSyn$colMeansTs)))
#Add moving mean
for (i in 370:900){
    range = (i-10):(i+10)
    points(i, median(RTdataSyn$colMeansTs[RTdataSyn$num%in%range])+0.001, col="blue",pch = 11, cex=2)
}
dev.off()
```

```{r}
AlldataSyn<-rbind(PROdata[PROdata$TypeOfSite=="syn",],RTdata[RTdata$TypeOfSite=="syn",])
plot(AlldataSyn$num[order(AlldataSyn$num)],AlldataSyn$colMeansTs[order(AlldataSyn$num)]+0.001,col=AlldataSyn$color[order(AlldataSyn$num)],pch=16,
     log="y",
     type="b", 
    ylim = c(0.001,0.06)    )#,ylim=c(0,max(RTdataNonSyn$colMeansTs)))
#Add line for AA 82-107 from Mayrose paper
lines(c(82*3,107*3),c(0.002,0.002),col=2,lwd=2)
lines(c(7*3,31*3),c(0.002,0.002),col=2,lwd=2)


```

## RT
RTdataNonSyn<-RTdata[RTdata$TypeOfSite=="nonsyn",]
plot(RTdataNonSyn$num[order(RTdataNonSyn$num)],RTdataNonSyn$colMeansTs[order(RTdataNonSyn$num)]+0.001,col=RTdataNonSyn$color[order(RTdataNonSyn$num)],pch=16,log="y",type="b")

#Add moving mean
for (i in 370:900){
    range = (i-20):(i+20)
    points(i, mean(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="blue",pch = 12, cex=3)
    points(i, median(RTdataNonSyn$colMeansTs[RTdataNonSyn$num%in%range]), col="green",pch = 11, cex=3)
}

i=370
    range = (i-20):(i+20)
    
```    



# Next steps 
Going to randomisations
- frequency-ordered figure: does the coloring scheme hold?
  
- position-ordered figure: does the pattern hold?  
    
###For protease, which are the most deleterious non-syn sites? 
To be done 
Interesting? why? for therapy or vaccine. 

###For protease, which are the most neutral non-syn sites? 
corresponding to the expected frequency based on sequence length * mutation rate? Frequency should flucatuate aroudn the expected value? 

#For protease, which are the most deleterious syn sites? 
To be done : green colored sites with low frequ
Interesting why?  HIV replication?





